File: apps/render-android-playground/app/src/main/java/com/max_mrtnv/render_android_playground/infrastructure/network/NetworkClient.kt
package com.max_mrtnv.render_android_playground.infrastructure.network

import com.google.gson.Gson
import com.max_mrtnv.render_android_playground.application.errors.ApplicationError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException

/**
 * Network client for making HTTP requests
 */
class NetworkClient(
    private val client: OkHttpClient = OkHttpClient(),
    private val gson: Gson = Gson()
) {
    suspend fun fetchData(url: String): String {
        return withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url(url)
                .build()
            
            try {
                val response = client.newCall(request).execute()
                
                if (!response.isSuccessful) {
                    throw ApplicationError.InvalidResponse("HTTP status code: ${response.code}")
                }
                
                response.body?.string() ?: throw ApplicationError.InvalidResponse("Empty response body")
            } catch (e: IOException) {
                throw ApplicationError.NetworkError(e.message ?: "Network request failed")
            } catch (e: ApplicationError) {
                throw e
            } catch (e: Exception) {
                throw ApplicationError.NetworkError(e.message ?: "Unknown network error")
            }
        }
    }
    
    suspend fun fetchJson(url: String): Map<String, Any?> {
        val data = fetchData(url)
        
        return try {
            // Use a more generic approach that handles LinkedTreeMap properly
            val result = gson.fromJson(data, Any::class.java)
            when (result) {
                is Map<*, *> -> result.mapKeys { it.key.toString() }.mapValues { it.value }
                else -> throw ApplicationError.ParsingError("JSON root is not an object")
            }
        } catch (e: Exception) {
            throw ApplicationError.ParsingError("Failed to parse JSON: ${e.message}")
        }
    }
}

File: apps/render-ios-playground/render-ios-playground/SDK/Core/Network/NetworkClient.swift
import Foundation

/// Network client for making HTTP requests
class NetworkClient {
    private let session: URLSession
    
    init(session: URLSession = .shared) {
        self.session = session
    }
    
    func fetchData(from url: URL) async throws -> Data {
        do {
            let (data, response) = try await session.data(from: url)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw ApplicationError.invalidResponse("Invalid response type")
            }
            
            guard httpResponse.statusCode == 200 else {
                throw ApplicationError.invalidResponse("HTTP status code: \(httpResponse.statusCode)")
            }
            
            return data
        } catch {
            if error is ApplicationError {
                throw error
            } else {
                throw ApplicationError.networkError(error.localizedDescription)
            }
        }
    }
    
    func fetchJSON(from url: URL) async throws -> [String: Any] {
        let data = try await fetchData(from: url)
        
        do {
            guard let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
                throw ApplicationError.parsingError("Response is not a valid JSON object")
            }
            return json
        } catch {
            throw ApplicationError.parsingError("Failed to parse JSON: \(error.localizedDescription)")
        }
    }
}
